packages/aws-rfdk/lib/core/lambdas/nodejs/version-provider/version-provider.ts

```diff
@@ -0,0 +1,318 @@
+   */
+  private readInstallersIndex(): any {
+    if (!this.indexFilePath) {
+      throw new Error('File path should be defined.');
+    }
+    if (!fs.existsSync(this.indexFilePath)) {
+      throw new Error(`File ${this.indexFilePath} was not found`);
+    }
+    const data = fs.readFileSync(this.indexFilePath, 'utf8');
+
+    // convert the response to a json object and return.
+    const json = JSON.parse(data);
+    return json;
+  }
+
+  private parseVersionString(versionString: string): RegExpExecArray | null {
+    return this.VALID_VERSION_REGEX.exec(versionString);
+  }
+
+  /**
+   * This method returns IVersionedUris for specific platform
+   *
+   * @param product
+   * @param productSection
+   * @param platform
+   * @param version
+   */
+  private async getUrisForPlatform(
+    product: Product,
+    productSection: any,
+    platform: Platform,
+    version?: string,
+  ): Promise<IVersionedUris | undefined> {
+    const versionString: string = version ? version : this.getLatestVersion(platform, productSection);
+
+    const requestedVersion = this.parseVersionString( versionString );
+
+    // Based on the requested version, fetches the latest patch and its installer file paths.
+    return this.getRequestedUriVersion(
+      requestedVersion,
+      productSection.versions,
+      platform,
+      product,
+    );
+  }
+
+  /**
+   * This method returns the latest version for specified platform.
+   *
+   * @param platform
+   * @param indexedVersionInfo
+   */
+  private getLatestVersion(platform: string, indexedVersionInfo: any): string {
+    const latestSection = indexedVersionInfo.latest;
+    if (!latestSection) {
+      throw new Error('Information about latest version can not be found');
+    }
+
+    const latestVersion = latestSection[platform];
+    if (!latestVersion) {
+      throw new Error(`Information about latest version for platform ${platform} can not be found`);
+    }
+
+    return latestVersion;
+  }
+
+  /**
+   * This method looks for the requested version (partial or complete) in the
+   * indexed version information. Based on the input, it iterates through all
+   * four numbers in the version string and compares the requested version
+   * with the indexed info.
+   * If any of the requested version number is missing, it fetches the latest
+   * (highest) available version for it.
+   *
+   * @param requestedVersion
+   * @param indexedVersionInfo
+   */
+  private getRequestedUriVersion(
+    requestedVersion: RegExpExecArray | null,
+    indexedVersionInfo: any,
+    platform: Platform,
+    product: Product,
+  ): IVersionedUris | undefined {
+
+    let versionMap = indexedVersionInfo;
+    const versionArray: string[] = [];
+
+    // iterate through all 4 major, minor, release and patch numbers,
+    // and get the matching version from the indexed version map.
+    for (let versionIndex = 0; versionIndex < 4; versionIndex++) {
+      let version: string;
+      if (null === requestedVersion
```

Comment:
can this be changed to using optional chaining something along the lines of

requestedVersion?.[versionIndex + 1] == null